Fix User leakage in Batch enrollment

melvinsoft opened the pull request:
We're experiencing problems with the Batch Enrollment section, that is causing small leaks between sites. But before, is important to understand how it works.

The Batch enrollment section is basically a TextArea in where you can enter a comma / line separated list of usernames or emails (yeah, all mixed in the same list), the batch enrollment will iterate over all of those users and do the following:
- If the user exists by username or email: Will be enrolled in course.
- If the user doesn't exists by username, it won't do nothing and notify in the UI that doesn't exists. 
- If the user doesn't exists by email, it will be pre-enrolled and an invitation will be sent. Pre-enrolled means that EnrollmentAllowed record is created, so at the time the user registers, it'll be automatically enrolled in the course.

In Tahoe we're having several issues, because we're emulating a multi tenant system:
1 - If a user is registered in any Tahoe site with a certain email, it can be automatically enrolled in other Tahoe site by email, this causes data leakage in reports, and also the user receives an invitation for a course in a site that cannot access.
2 - If you paste random username like `carlos` and by any chance a user with that username exists in another Tahoe site, it will be enrolled in the course, causing the same problem described above. 

The tricky part of this issue, is that we cannot remove the functionality of invite non registered users by email, because is the charm of the feature, invite user to register and enroll in a course.

The PR solves most of those issues, but simply check if the user exists inside the organization, if is searched by username and not exists in the site, even if exists in another site, it simply will return a not exists. If the user is searched by email, despite if exists in another site or not, but don't exists in the current one, it'll be invited by email.

The edge case is, if the user exists in another site, it won't be able to sign up, but that's a different problem that will be tackled here: https://github.com/appsembler/tech-design-proposals/pull/8

johnbaldwin commented on lms/djangoapps/instructor/enrollment.py:
```diff
@@ -26,6 +26,7 @@
 from lms.djangoapps.grades.signals.signals import PROBLEM_RAW_SCORE_CHANGED
 from openedx.core.djangoapps.lang_pref import LANGUAGE_KEY
 from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
+from openedx.core.djangoapps.theming.helpers import get_current_site
```
This method returns None if `get_current_request` returns None. See here: https://github.com/appsembler/edx-platform/blob/appsembler/tahoe/master/openedx/core/djangoapps/theming/helpers.py#L175-L185

Which means the call below using it can raise an exception if the request is not cached in the RequestCache. I'd prefer that we explicitly pass the site object to the `EmailEnrollmentState` constructor

OmarIthawi replied:
@johnbaldwin as far as I know this method is supposed to always work, unless it's called in a non-web view e.g. in a Celery task.

It should be safe to use as far as I know. Is there anything else I don't know?

johnbaldwin replied:
@OmarIthawi As long as you are confident it is *always*, then I'll skip this one. However, it would be good to test this that the code fails how we want if something in the future changes the contract of *always*.

melvinsoft replied:
@johnbaldwin I searched for all the call to this method, as we're safe using it, but I agree with you, we need to implement tests to be sure.

melvinsoft replied:
@johnbaldwin Thanks! I could use your help next sprint to implement tests here, in the meantime, I'm going to close this comment, and start documenting the tests we need to implement.

johnbaldwin commented on lms/djangoapps/instructor/enrollment.py:
```diff
@@ -52,9 +53,12 @@ def __init__(self, course_id, email):
         # handled at a higher level (i.e. in enroll_email).  Besides, this
         # class creates readonly objects.
-        exists_user = User.objects.filter(email=email).exists()
+        organization = get_current_site().organizations.first()
```
This can raise an exception if the `RequestCache` class does not cache the current request object. If we are not in a view class that has a request object, we should instead pass the `site` instance to classes and methods that need it. This both makes it explicit and helps makes tests more robust.

What it even better is to centralize site handling methods into one module. for an example, see the `sites` module in the appsembler api: https://github.com/appsembler/edx-platform/blob/appsembler/tahoe/master/openedx/core/djangoapps/appsembler/api/sites.py

Making a central point for setting/getting site specific data lets us both reuse code and do focused unit testing for site isolation

OmarIthawi replied:
I like the refactoring idea you're suggesting @johnbaldwin, `.organizations.first()` is being used a lot and it should be placed in a function instead of copying the pattern all over tahoe.

johnbaldwin commented on lms/djangoapps/instructor/enrollment.py:
```diff
@@ -52,9 +53,12 @@ def __init__(self, course_id, email):
         # class creates readonly objects.
-        exists_user = User.objects.filter(email=email).exists()
+        organization = get_current_site().organizations.first()
+        exists_user = organization.userorganizationmapping_set.filter(user__email=email).exists()
```
so this call could be `exists_user = sites.site_has_user(site=site, user_email=email)`

johnbaldwin replied:
And it looks like `EmailEnrollmentState` is only used in this module (`instructor.enrollments`), so when we add `site` as a constructor param, we only have to change it in this module and the `test_enrollment` module

```
./lms/djangoapps/instructor/tests/test_enrollment.py:26:    EmailEnrollmentState,
./lms/djangoapps/instructor/tests/test_enrollment.py:62:        ees = EmailEnrollmentState(self.course_key, eobjs.email)
./lms/djangoapps/instructor/tests/test_enrollment.py:94:        before = EmailEnrollmentState(self.course_key, eobjs.email)
./lms/djangoapps/instructor/tests/test_enrollment.py:103:        after = EmailEnrollmentState(self.course_key, eobjs.email)
./lms/djangoapps/instructor/tests/test_enrollment.py:597:class SettableEnrollmentState(EmailEnrollmentState):
./lms/djangoapps/instructor/enrollment.py:49:class EmailEnrollmentState(object):
./lms/djangoapps/instructor/enrollment.py:126:    returns two EmailEnrollmentState's
./lms/djangoapps/instructor/enrollment.py:129:    previous_state = EmailEnrollmentState(course_id, student_email)
./lms/djangoapps/instructor/enrollment.py:162:    after_state = EmailEnrollmentState(course_id, student_email)
./lms/djangoapps/instructor/enrollment.py:176:    returns two EmailEnrollmentState's
./lms/djangoapps/instructor/enrollment.py:179:    previous_state = EmailEnrollmentState(course_id, student_email)
./lms/djangoapps/instructor/enrollment.py:196:    after_state = EmailEnrollmentState(course_id, student_email)
```

OmarIthawi replied:
@johnbaldwin I'd be really careful not to change so many files. It's very easy that we'd forget to migrate future uses when upgrading to a newer release.

Git conflicts can be really unproductive to resolve.

If we really want to be careful, we can still use something like this function:

https://github.com/appsembler/edx-platform/blob/64555c60c043acb00ff68f508fad73d09691fc12/openedx/core/djangoapps/request_cache/__init__.py#L63-L72 

Which addresses the issues of having the site returned as `None`.

johnbaldwin replied:
Reading that function, a dummy site is in the fake request is returned. In this case, the user should be rejected. We should test that this behaves how we want

johnbaldwin commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -3239,7 +3241,9 @@ def build_row_errors(key, _user, row_count):
-                user = get_user_by_username_or_email(user)
+                # Appsembler Specific: We call our custom method insted the default one,
+                # to make sure the user is get inside the org.
+                user = get_user_by_username_or_email_inside_organization(user)
```
Following on my comments above, I need to look at this more carefully, but we might be able to just bypass the "user belongs to organization" checks in the `EmailEnrollmentState` by doing the check before `EmailEnrollmentState` is even called. 

only `instructor.enrollment.enroll_email` and `instructor.enrollment.unenroll_email` instantiate `EmailEnrollmentState` objects. Therefore, if we check if the user already exists in a site (or organization) in either the `enroll_email` or `unenroll_email` functions or in the functions in `instructor.views.api` that call `enroll_email` and `unenroll_email`, we don't need to do site specific handling in the `EmailEnrollmentState` constructor.

However, all this will require reworking code in `instructor.views.api` to explicitly handle `site` objects, which may be more work than we want to do for this bug fix.

johnbaldwin replied:
Please disregard the above comment. @OmarIthawi is right, this touches too much code for the value

melvinsoft replied:
@johnbaldwin Yes, right now, we don't want to introduce so many changes in the platform, since it makes code diffs really difficult later. But I agree it should be the way.

johnbaldwin commented:
@melvinsoft I made a bunch of comments suggesting an implementation approach that explicitly passes the current `site` object instead of relying on the `RequestCache` to contain the `request` object that defines the site.  Probably more work than we want to do.

After thinking about it _after I typed up my comments_ I decided to leave my comments in place just so you see what I'm thinking based on my efforts to try to make site isolation safer and easier to test

Probably the most expedient things to do instead are

A) Be clear in how the code works for when `RequestCache` does not have a valid request
B) Add test to handle when this call fails: https://github.com/appsembler/edx-platform/blob/a3c94894a0dbb074e8c1b948f3e0fbaacd354b11/lms/djangoapps/instructor/enrollment.py#L56

C) Add test to run the code for a user that doesn't belong to the site/organization
D) Add test to run the code for a user that DOES belong to the site/organization

For examples, please check out my tests for the Tahoe API: https://github.com/appsembler/edx-platform/tree/appsembler/tahoe/master/openedx/core/djangoapps/appsembler/api/tests

melvinsoft commented:
@johnbaldwin @OmarIthawi Thank you so much for your reviews. I've implemented the first round of changes, basically refactoring some code into functions and adding a feature flag to control which function we call as @OmarIthawi requested.

After chatting with @abeals, since this bug is pressuring us, test may come in a second iteration next sprint.

johnbaldwin commented on lms/djangoapps/instructor/sites.py:
```diff
@@ -0,0 +1,27 @@
+"""
```
@melvinsoft We may want to move this module out of `instructor` to a common place. Since our site checking depends on Appsembler's fork of `organizations`, how about moving it to `openedx.core.djangoapps.appsembler.sites.api`? The `api` module seems reasonable to me

johnbaldwin commented on lms/djangoapps/instructor/sites.py:
```diff
@@ -0,0 +1,27 @@
+
+def user_exists_in_organization(user_email, organization):
+    """
+    TODO
```
What are the `TODO` statements for?

melvinsoft replied:
Just to write comments for the functions.

johnbaldwin commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -3124,7 +3125,12 @@ def get_student(username_or_email, course_key):
-        student = get_user_by_username_or_email(username_or_email)
+        # Appsembler Specific: We call our custom method insted the default one,
+        # to make sure the user is get inside the org.
+        if settings.FEATURES.get('TAHOE_MULTITENANT_BULK_ENROLLMENT', False):
```
Cool, thanks!

melvinsoft replied:
👍

johnbaldwin commented on lms/djangoapps/instructor/views/tools.py:
```diff
@@ -67,7 +67,9 @@ def get_student_from_identifier(unique_student_identifier):
-    return get_user_by_username_or_email(unique_student_identifier)
+    # Appsembler Specific: We call our custom method insted the default one,
+    # to make sure the user is get inside the org.
+    return get_user_by_username_or_email_inside_organization(unique_student_identifier)
```
@melvinsoft Do we need to add a feature flag here too?

melvinsoft replied:
Yes, I missed it, thanks! 👍

melvinsoft commented:
@johnbaldwin I fixed some PEP8 issues, but I think the checks are failing because the files we changed, have PEP8 errors from before, from example the file `common/djangoapps/student/models.py` is 2962 lines long, and is showing a hundreds of PEP8 errors for lines longer than 79 chars, fixing all of those is out of scope of this PR.

Please give it another round of review and approve it if it's meeting the criteria, according our chat yesterday.

melvinsoft commented:
@thraxil Thanks for addressing the PEP8 issues! 

@johnbaldwin Another ping here for a final review and approval. Thanks!

johnbaldwin commented on lms/djangoapps/instructor/enrollment.py:
```diff
@@ -52,9 +58,13 @@ def __init__(self, course_id, email):
-        exists_user = User.objects.filter(email=email).exists()
+        site = get_current_site()
+        organization = get_organization_for_site(site)
+        exists_user = user_exists_in_organization(email, organization)
```
@melvinsoft why do a user check instead of just a single call to try to get the user and use `if user` instead of `if exists_user`? Is seems to make an extra call that is not needed.

melvinsoft replied:
@johnbaldwin Thanks for asking this, this is edX code, no ours, we trying to fix the issue with the less code changes as possible, since makes future merges way more difficult. I don't like the design, but I choose to stick with the code as much as I can.

johnbaldwin replied:
OK, Thanks
